genesis/engine/coupler.py

```diff
@@ -952,6 +976,102 @@ def fem_floor_detection(self, f: ti.i32):
+
+        # Compute data for each contact pair
+        for i_c in range(self.n_fem_floor_contact_pairs[None]):
+            pair = self.fem_floor_contact_pairs[i_c]
+            self.fem_floor_contact_pairs[i_c].active = 1  # mark the contact pair as active
+            i_b = pair.batch_idx
+            i_e = pair.geom_idx
+            intersection_code = pair.intersection_code
+            distance = pair.distance
+            intersected_edges = ti.static(self.kMarchingTetsEdgeTable)[intersection_code]
+            tet_vertices = ti.Matrix.zero(gs.ti_float, 3, 4)  # 4 vertices
+            tet_pressures = ti.Vector.zero(gs.ti_float, 4)  # pressures at the vertices
+
+            for i in ti.static(range(4)):
+                i_v = fem_solver.elements_i[i_e].el2v[i]
+                tet_vertices[:, i] = fem_solver.elements_v[f + 1, i_v, i_b].pos
+                tet_pressures[i] = self.fem_pressure[i_v]
+
+            polygon_vertices = ti.Matrix.zero(gs.ti_float, 3, 4)  # 3 or 4 vertices
+            total_area = gs.EPS  # avoid division by zero
+            total_area_weighted_centroid = ti.Vector([0.0, 0.0, 0.0])
+            for i in range(4):
+                if intersected_edges[i] >= 0:
+                    edge = ti.static(self.kTetEdges)[intersected_edges[i]]
+                    pos_v0 = tet_vertices[:, edge[0]]
+                    pos_v1 = tet_vertices[:, edge[1]]
+                    d_v0 = distance[edge[0]]
+                    d_v1 = distance[edge[1]]
+                    t = d_v0 / (d_v0 - d_v1)
+                    polygon_vertices[:, i] = pos_v0 + t * (pos_v1 - pos_v0)
+
+                    # Compute tirangle area and centroid
+                    if i >= 2:
+                        e1 = polygon_vertices[:, i - 1] - polygon_vertices[:, 0]
+                        e2 = polygon_vertices[:, i] - polygon_vertices[:, 0]
+                        area = 0.5 * e1.cross(e2).norm()
+                        total_area += area
+                        total_area_weighted_centroid += (
+                            area * (polygon_vertices[:, 0] + polygon_vertices[:, i - 1] + polygon_vertices[:, i]) / 3.0
+                        )
+
+            centroid = total_area_weighted_centroid / total_area
+
+            # Compute barycentric coordinates
+            barycentric = tet_barycentric(centroid, tet_vertices)
+            pressure = (
+                barycentric[0] * tet_pressures[0]
+                + barycentric[1] * tet_pressures[1]
+                + barycentric[2] * tet_pressures[2]
+                + barycentric[3] * tet_pressures[3]
+            )
+            self.fem_floor_contact_pairs[i_c].barycentric = barycentric
+
+            C = ti.static(1.0e8)
+            deformable_g = C
+            rigid_g = self.fem_pressure_gradient[i_b, i_e].z
+            # TODO A better way to handle corner cases where pressure and pressure gradient are ill defined
+            if total_area < gs.EPS or rigid_g < gs.EPS:
```

Comment:
`rigid_g < gs.EPS` should not be an issue mathematically at least (see following comment).